### DISCUSSION TITLE
Utilizing Zynq Timestamping to Improve Compatibility of PlutoSDR with LTESniffer

### DISCUSSION THREAD

**Discussion by apex-jpg** [srsran/zynq_timestamping#60] [General]
Hello!

I'm currently involved in a project aimed at LTE signal analysis using the [LTESniffer tool](https://github.com/SysSec-KAIST/LTESniffer) and an ADALM Pluto SDR device. I am exploring the potential of zynq timestamping implementation to tackle certain issues I am facing.

### About LTESniffer

LTESniffer is a versatile, open-source tool built on top of [FALCON ](https://github.com/falkenber9/falcon) and srsRAN. It provides a range of capabilities including uplink and downlink sniffing, image mapping, and information extraction.

### Encountered Issues

I've been experiencing [issues](https://github.com/SysSec-KAIST/LTESniffer/issues/31#issuecomment-1630450289) with uplink and downlink sniffing. 

According to the authors of LTESniffer, this might stem from a hardware compatibility issue - the requirement of an SDR device to support 2 clock sources for 2 RX antennas. This is not an inherent feature of the PlutoSDR, which I am using.

I came across this repo while I was stumbling around for possible solutions, and have been curious about its precise sample alignment in SDR implementations through internal FPGA clock ticks counting, which might potentially provide the compatibility I am seeking.

For clarity, I am conducting this project responsibly within a controlled environment (utilizing a Faraday cage) to ensure compliance with relevant licensing and regulations.

I am seeking your insights on the following:
1. Could implementing "zynq_timestamping" potentially enable my PlutoSDR to be fully compatible with LTESniffer, especially considering the need for 2 clock sources for 2 RX antennas?
2. Are there any known limitations or potential challenges that I should be aware of when attempting to integrate "zynq_timestamping" with PlutoSDR and LTESniffer?
3. Could it be possible to overcome the hardware limitations by modifying the codebase of srsRAN to enhance PlutoSDR's compatibility?

I appreciate you reading through this! Your input will be invaluable to my project.

Thanks!

**Comment by ofontbach** [comment]
Hi @apex-jpg,

Thanks for reposting the question that you sent to the srsLTE user lists here, so it might reach the community better. I'll also repost my answer to you here with the same objective.

The Zynq Timestamping solution was designed with the objective of being as generic as possible, but to be fair we did optimize it for medium/large Zynq devices, having in mind to enable accelerating big parts of the PHY on the FPGA too. This means that we never really took the time to fully optimize it for the ADALM-PLUTO. If you dive into the discussions and issues sections, you will see that different users have tried to use the ADALM-PLUTO with different levels of success. Unfortunately, it seems that the USB 2.0 in the ADALM-PLUTO has issues keeping a stable 1.92 MSPS exchange with the host when using regular IIO. Two main solutions can be applied to overcome that: i) using large iio_buffers, as described [in this thread](https://github.com/srsran/zynq_timestamping/issues/35#issuecomment-1452218133) or ii) running the full srsRAN stack in the Zynq device. Unfortunately, using large buffers is not mixing well with LTE's HARQ requirements. Regarding the latter, while the bandwidth between the embedded ARM and the FPGA is no longer an issue, the truth is that without offloading part of the PHY processing to the FPGA, you will only be able to run applications with limited performance requirements. 

This being said, I can see three main options on your end:

1) Use a different Zynq-based SDR. For instance, the AntSDR, which basically is an ADALM-PLUTO with Ethernet interfacing that can be found at a similar low-level price point. We have successfully tested it with srsRAN, as detailed [here](https://github.com/srsran/zynq_timestamping/discussions/27#discussioncomment-4424302). If you take this route, the Zynq Timestamping solution should be plug-and-play for you. Other (more expensive) Zynq boards have also been extensively used with srsRAN, as detailed in the documentation, although this use case is mostly used for fully-embedded implementations (i.e., the srsRAN stack fully runs in the Zynq).

2) Put some development/tuning effort to make the Zynq Timestamping work more stably with the ADALM-PLUTO - as mentioned before, you can find abundant details of what other hackers have tried and the results they got in the pointed discussions and issue links. That sure would be an excellent contribution ;)

3) Give a go to the zynq timestamping-flavoured solution provided by @pgreenland, as described in [this thread](https://github.com/srsran/zynq_timestamping/issues/35#issuecomment-1577496707). Honestly, I'm not sure how direct this solution will be for you or if you'll find other bottlenecks, but it has been specifically tailored for the ADALM-PLUTO.

Best regards,
Oriol

**Comment by lucaaamaral** [comment]
Hey, I am wondering how is the current status of the ZynqSDR pluto support with srsRAN, can I use it as a fully embedded RAN SDR? Is it compatible with both 4G and 5G?

**Reply by ofontbach** [reply]
Hi @lucaaamaral,

We've never tested it, but it seems quite similar to the AntSDR. Hopefully, you'd be able to reuse the solution with minimal effort. Currently, the zynq timestamping solution only supports its utilization with the 4G codebase.

Cheers

**Reply by lucaaamaral** [reply]
Thank you for the response @ofontbach . Do you know what is missing for using with 5G? I might spare some time to implement myself or assign to somebody else.
